### PR TITLE
feat: add action to remove component dependencies on deletion

### PR DESCRIPTION
**Pull request by fuka-michal**
Implements a right-click context menu and file watcher that automatically removes all associated dependencies when a component is deleted locally. This ensures a clean environment and prevents orphaned dependencies in the project structure.

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -31,6 +31,12 @@ import AccountCommands = require('./commands/AccountCommands');
 import EnvironmentCommands = require('./commands/EnvironmentCommands');
 import PublicCommands = require('./commands/PublicCommands');
 import { telemetryReporter, sendTelemetry, startAppInsights } from './utils/telemetry';
+import { getMakecomappJson, getMakecomappRootDir } from './local-development/makecomappjson';
+import { AppComponentType, AppComponentTypes } from './types/app-component-type.types';
+import { deleteLocalComponent } from './local-development/delete-local-component';
+import { catchError } from './error-handling';
+import { Uri } from 'vscode';
```
Comment (ma-zal) [inline src/extension.ts:38]: ... Already imported on the top by "include * as vscode..."

```suggestion
```

File: src/extension.ts [hunk 2]
```diff
@@ -232,6 +238,52 @@ export async function activate(context: vscode.ExtensionContext) {
...
+		const matchComponentPath = path.match(regex);
+
+		if (matchComponentPath) {
+			const [, componentType, componentName] = matchComponentPath;
+			return { componentType: componentType as AppComponentType, componentName };
+		}
+		return null;
+	}
+
+	async function onFileDeleted(uri: Uri) {
```
Comment (ma-zal) [inline src/extension.ts:259]: Uses the already existing import:

```suggestion
	async function onFileDeleted(uri: vscode.Uri) {
```

File: src/extension.ts [hunk 3]
```diff
@@ -232,6 +238,52 @@ export async function activate(context: vscode.ExtensionContext) {
...
 	);
 
+	// Component deletion context menu.
+	vscode.commands.registerCommand(
+		'apps-sdk.local-dev.delete-local-component',
+		// Delete folder will trigger file watcher
+		catchError('Delete local component', removeRecursively),
+	);
+
+	function parseComponentPath(path: string) {
```
Comment (ma-zal) [inline src/extension.ts:248]:
```suggestion
	function parseComponentPath(path: string): { componentType: AppComponentType; componentName: string } | null {
```

File: src/local-development/helpers/component-id-mapping-helper.ts [hunk 4]
```diff
@@ -98,4 +127,55 @@ export class ComponentIdMappingHelper {
...
+			if (originInMakecomappJson.idMapping?.[componentType]) {
+				originInMakecomappJson.idMapping[componentType] = originInMakecomappJson.idMapping[componentType].filter(filterCondition);
+			}
+			// Modify origin object that
+			if (this.origin.idMapping?.[componentType]) {
+				this.origin.idMapping[componentType] = this.origin.idMapping[componentType].filter(filterCondition);
+			}
+
+		}
+		return originInMakecomappJson;
```
Comment (ma-zal) [inline src/local-development/helpers/component-id-mapping-helper.ts:179]: Returning values if not used anywhere in implementation. Consider remove the `return`.

```suggestion
```

File: src/local-development/helpers/fs.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+	} catch (error) {
+		if (error instanceof vscode.FileSystemError && error.code === 'FileNotFound') {
+			return false; // File does not exist
+		}
+		// Handle other errors if necessary
+		throw error;
+	}
+}
+
+export async function removeRecursively(uri: Uri) {
```
Comment (ma-zal) [inline src/local-development/helpers/fs.ts:17]:
```suggestion
export async function removeRecursively(uri: vscode.Uri) {
```

File: src/local-development/helpers/fs.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import * as vscode from 'vscode';
+import { Uri } from 'vscode';
```
Comment (ma-zal) [inline src/local-development/helpers/fs.ts:2]: Not needed to import Uri separatly.

```suggestion
import * as vscode from 'vscode';
```

File: src/local-development/helpers/fs.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+			return false; // File does not exist
+		}
+		// Handle other errors if necessary
+		throw error;
+	}
+}
+
+export async function removeRecursively(uri: Uri) {
+	await vscode.workspace.fs.delete(uri, { recursive: true });
+}
```
Comment (ma-zal) [inline src/local-development/helpers/fs.ts:19]: Consider if it is useful to wrap one-line function call by another function here.

File: src/local-development/align-components-mapping.ts [hunk 8]
```diff
@@ -52,13 +53,32 @@ export async function alignComponentsMapping(
 		componentName: string;
 		componentMetadata: AppComponentMetadata;
 	}[] = [];
+	/**
+	 * Deleted locally: Missing in `makecomappJson`, but existing in `remoteComponents` and 'mapping'.
```
Comment (ma-zal) [inline src/local-development/align-components-mapping.ts:57]:
```suggestion
	 * Deleted locally: Missing in local filesystem, but existing in `remoteComponents` and 'mapping'.
```

File: src/local-development/delete-local-component.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import { AppComponentType } from '../types/app-component-type.types';
+import { log } from '../output-channel';
+import { removeModuleFromGroups } from './groups-json';
+import vscode from 'vscode';
```
Comment (ma-zal) [inline src/local-development/delete-local-component.ts:4]: Import everywhere else is "* as":

```suggestion
import * as vscode from 'vscode';
```

File: src/local-development/helpers/fs.ts (hunk 7 above)
Comment (fuka-michal) [reply src/local-development/helpers/fs.ts:19]: renamed to delete with options, wrapped due to feature gh actions

File: src/local-development/delete-origin-component.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import { AppComponentType } from '../types/app-component-type.types';
+import { log } from '../output-channel';
+import { requestMakeApi } from '../utils/request-api-make';
+import { AxiosRequestConfig } from 'axios';
+import { version as ExtensionVersion } from '../Meta';
+import { LocalAppOriginWithSecret } from './types/makecomapp.types';
+import { getComponentApiUrl } from './helpers/api-url';
+
+export async function deleteOriginComponent(origin: LocalAppOriginWithSecret, componentType: AppComponentType, remoteComponentName: string) {
+	log('info', `Deleting origin ${componentType} '${remoteComponentName}'`);
```
Comment (ma-zal) [inline src/local-development/delete-origin-component.ts:10]:
```suggestion
	log('info', `Deleting remote ${componentType} '${remoteComponentName}' in origin "${origin.label || origin.appId}"`);
```

File: src/local-development/delete-origin-component.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+	const axiosConfig: AxiosRequestConfig = {
+		url: getComponentApiUrl({ componentType: componentType, remoteComponentName, origin }),
+		method: 'DELETE',
+		headers: {
+			Authorization: 'Token ' + origin.apikey,
+			'imt-vsce-localmode': 'true',
+			'imt-apps-sdk-version': ExtensionVersion,
+		},
+	};
+	log('info', `Deleted origin ${componentType} '${remoteComponentName}'`);
```
Comment (ma-zal) [inline src/local-development/delete-origin-component.ts:21]:
```suggestion
	log('info', `Deleted remote ${componentType} '${remoteComponentName}'`);
```

File: src/local-development/pull.ts [hunk 12]
```diff
@@ -49,6 +54,13 @@ export async function pullAllComponents(
 	origin: LocalAppOriginWithSecret,
 	newRemoteComponentResolution: 'askUser' | 'cloneAsNew',
 ): Promise<void> {
+
+	// Pulling when there are undeployed changes. Override them and align the state with the origin.
+	const makecomappJson = await getMakecomappJson(localAppRootdir);
+	const componentIdMappingHelper = new ComponentIdMappingHelper(makecomappJson, origin);
+	componentIdMappingHelper.filterMappingItems(item => !item?.localDeleted);
```
Comment (ma-zal) [inline src/local-development/pull.ts:61]:
```suggestion
	componentIdMappingHelper.filterMappingItems(item => !item?.localDeleted); // This is tricky command: It updates `this.makecomappJson, which === `const makecomappJson`.
````

File: src/local-development/align-components-mapping.ts [hunk 13]
```diff
@@ -109,6 +129,18 @@ export async function alignComponentsMapping(
...
+		const deletedLocallyInSpecificComponentType = deletedLocally.filter(
+			(component) => component.componentType === componentType,
+		);
+
+		// Resolve locally deleted components
+		let deletedLocallyComponent;
+		while ((deletedLocallyComponent = deletedLocallyInSpecificComponentType.shift()!)) {
+			await deleteOriginComponent(origin, componentType, deletedLocallyComponent.componentName);
+			const mappingHelper = new ComponentIdMappingHelper(makecomappJson, origin);
+			mappingHelper.removeByRemoteName(componentType, deletedLocallyComponent.componentName);
```
Comment (ma-zal) [inline src/local-development/align-components-mapping.ts:141]:
```suggestion
			mappingHelper.removeByRemoteName(componentType, deletedLocallyComponent.componentName); // this is tricky command: It updates `this.makecomappJson`, which === `makecomappJson`
```

**Review by ma-zal — APPROVED** [review]

**Pull request merged**